File: src/NotificationsService/Service/Queries/GetNotificationCountQueryHandler.cs
using System.Data;
using Dapper;
using MediatR;
using NotificationsService.Database.Queries;
using NotificationsService.Service.Api.Queries;

namespace NotificationsService.Service.Queries;

/// <summary>
/// A handler class for the GetNotificationCountQuery query.
/// </summary>
public sealed class GetNotificationCountQueryHandler : IRequestHandler<GetNotificationCountQuery, int>
{
    private readonly IDbConnection _connection;

    public GetNotificationCountQueryHandler(IDbConnection connection)
    {
        _connection = connection;
    }

    public async Task<int> Handle(GetNotificationCountQuery request, CancellationToken cancellationToken)
    {
        return await _connection.QuerySingleAsync<int>(
            SqlQueries.GetNotificationsCount,
            new { request.UserId }
        );
    }
}

File: src/NotificationService/Service/Queries/Handlers/GetNotificationCountQueryHandler.fs
namespace NotificationService.Service.Queries.Handlers

open MediatR
open NotificationService.Database.Api
open NotificationService.Service.Queries

/// A handler class for the GetNotificationCountQuery query.
type GetNotificationCountQueryHandler(repository: INotificationRepository) =
    interface IRequestHandler<GetNotificationCountQuery, int> with
        member this.Handle(request, cancellationToken) =
            task {
                if cancellationToken.IsCancellationRequested then
                    return 0
                else
                    return! repository.GetNotificationsCount(request.UserId)
            }
